src/librbd/io/ImageRequest.cc

```diff
@@ -34,37 +36,41 @@ namespace {
 template <typename I>
 struct C_UpdateTimestamp : public Context {
 public:
-  I* m_image_ctx;
-  AioCompletion* m_aio_comp;
-  bool modify; //if modify set to 'true', modify timestamp is updated, access timestamp otherwise
+  I& m_image_ctx;
+  bool m_modify; // if modify set to 'true', modify timestamp is updated,
+                 // access timestamp otherwise
+  AsyncOperation m_async_op;
 
-  C_UpdateTimestamp(I* ictx, AioCompletion *c, bool m)  
-    : m_image_ctx(ictx), m_aio_comp(c), modify(m) {
-    m_aio_comp->add_request();
+  C_UpdateTimestamp(I& ictx, bool m) : m_image_ctx(ictx), m_modify(m) {
+    m_async_op.start_op(*get_image_ctx(&m_image_ctx));
+  }
+  ~C_UpdateTimestamp() override {
+    m_async_op.finish_op();
   }
 
   void send() {
     librados::ObjectWriteOperation op;
-    if(modify)
+    if(m_modify) {
```

Comment:
nit: also fix missing whitespace after `if` while you are here?

Reply:
Done